src/admin/index.js

```diff
@@ -595,6 +597,55 @@ module.exports = ({
     }
   }
 
+  /**
+   * Describe cluster
+   *
+   * @return {Promise<ClusterMetadata>}
+   *
+   * @typedef {Object} ClusterMetadata
+   * @property {Array<Broker>} brokers
+   * @property {Number} controller Current controller id. Returns null if unknown.
+   * @property {String} clusterId
+   *
+   * @typedef {Object} Broker
+   * @property {Number} nodeId
+   * @property {String} host
+   * @property {Number} port
+   * @property {boolean} connected
+   */
+  const describeCluster = async () => {
+    const isConnected = async nodeId => {
+      try {
+        const broker = await cluster.brokerPool.findBroker({ nodeId, connect: false })
+        return broker.isConnected()
+      } catch (error) {
+        if (error.name === 'KafkaJSBrokerNotFound') {
+          return false
+        }
+
+        throw error
+      }
+    }
+
+    const { brokers: nodes, clusterId, controllerId } = await cluster.metadata({ topics: [] })
+    const brokers = await Promise.all(
+      nodes.map(async ({ nodeId, host, port }) => ({
+        nodeId,
+        host,
+        port,
+        connected: await isConnected(nodeId),
+      }))
+    )
```

Comment:
I wonder what the consequences of querying the connection state asynchronously are. Technically, you could get stale `connected` parameters for some brokers, as some might resolve faster than others. That _could_ be totally fine, but I'm not sure what the use case for describing the cluster is and using these flags is.

More generally speaking, is the connection state query asynchronous because we're trying to wait for state transitions to resolve (like `connecting -> connected`)? Is that appropriate to the use case?

Reply:
> is the connection state query asynchronous because we're trying to wait for state transitions to resolve (like connecting -> connected)?

No, the connection state query could technically be synchronous, because we are just looking at brokers that are already in memory, and then checking one of their properties to see if it is already connected. The only reason it is asynchronous is because it's using `Brokerpool.findBroker`, which normally also connects the broker if it is not already connected. For this use-case, we're passing an option that make it not connect the broker, so it's not really awaiting anything.

I could migrate the current `findBroker` to just... You know, _find_ the broker, and then create a new `findAndConnectBroker` or something that delegates to `findBroker` and then connects. That way `findBroker` can be completely synchronous.

Reply:
One problem for this particular use-case is that there's no way to know when the connection state changes, other than to poll `describeCluster`. That's really unrelated to this PR, but it could be solved with instrumentation events that are emitted on broker connect and disconnect.

Reply:
Ahh, for some reason I thought `broker.isConnected()` returned a `Promise` as well. That doesn't mean `findBroker` also creating a connection isn't an issue from a monitoring standpoint.

Taking a step back, I'm wondering what one would learn from a 'connected' flag when describing an cluster through the Admin? Surely, it's describing whether that Admin is connected through the broker at that instant, but is that relevant information? You'd probably want to know whether your consumers / producers are in a monitoring use case, but if we create and connect an Admin to describe the cluster, yeah, we're going to be connected to the brokers. I guess, it can show whether the brokers in the cluster are reachable?

Reply:
Well, you're connecting an admin client, so at least one broker is going to be connected, but you won't necessarily be connected to the other brokers in the cluster. We just get all the brokers from metadata, but we don't connect to them from the admin client.

To be honest, I don't really see what problem is solved by that either, but it is a requirement for #647, and it is simple enough to provide.

Reply:
While we _can_ add it easily enough, if the semantics don't really make sense for any use case of it, we're better of cutting it, as it's just waiting to be misinterpreted by someone.

> Taking a step back, I'm wondering what one would learn from a 'connected' flag when describing an cluster through the Admin? Surely, it's describing whether that Admin is connected through the broker at that instant, but is that relevant information? You'd probably want to know whether your consumers / producers are in a monitoring use case, but if we create and connect an Admin to describe the cluster, yeah, we're going to be connected to the brokers. I guess, it can show whether the brokers in the cluster are reachable?

@jlandersen you might hold the key to this and be able to enlighten us all 😊. Could you shine some light on why the `connected` flag would be useful for the use case of #647?

Reply:
I agree. Although I don't mind the inclusion of it, I would rather not introduce it if there isn't a good reason for it.

Reply:
I don't have a good reason to include a connected flag. I would leave it out as well.
My extension does show in the broker list whether connected to a specific broker, but besides that, there is no real reason for it. Will probably just remove that.

Reply:
Removed in 1ca7f25